File: dc-verifier/crates/dc-adapter-fastapi/python/dc_adapter_fastapi/extractor.py
"""
Извлекает метаданные из FastAPI приложения без запуска сервера.
Использует introspection API FastAPI.
"""
import importlib.util
import inspect
import sys
import logging
from pathlib import Path
from typing import Optional
from fastapi import FastAPI
from fastapi.routing import APIRoute

logger = logging.getLogger(__name__)


class FastApiExtractor:
    """Извлекает информацию о FastAPI приложении"""
    
    def __init__(self, app_path: str):
        self.app_path = Path(app_path)
        self.app: Optional[FastAPI] = None
    
    def load_app(self) -> FastAPI:
        """Загружает FastAPI приложение из файла"""
        if self.app is not None:
            return self.app
        
        # Динамический импорт модуля
        spec = importlib.util.spec_from_file_location("app", self.app_path)
        if spec is None or spec.loader is None:
            raise ValueError(f"Cannot load module from {self.app_path}")
        
        module = importlib.util.module_from_spec(spec)
        # Добавляем модуль в sys.modules для корректной работы относительных импортов
        sys.modules[spec.name] = module
        spec.loader.exec_module(module)
        
        # Ищем app в модуле
        if hasattr(module, "app") and isinstance(module.app, FastAPI):
            self.app = module.app
        else:
            raise ValueError(f"No FastAPI app found in {self.app_path}")
        
        return self.app
    
    def extract_routes(self) -> list[dict]:
        """Извлекает все routes из FastAPI приложения"""
        if self.app is None:
            self.load_app()
        
        routes = []
        for route in self.app.routes:
            if isinstance(route, APIRoute):
                # Обрабатываем возможные исключения при получении информации о файле
                try:
                    handler_file = inspect.getfile(route.endpoint)
                    handler_line = inspect.getsourcelines(route.endpoint)[1]
                except (TypeError, OSError) as e:
                    handler_file = None
                    handler_line = None
                    logger.warning(
                        f"Не удалось получить информацию о файле для {route.endpoint}: {e}"
                    )
                
                # Сохраняем все HTTP методы без мутации route.methods
                methods = list(route.methods) if route.methods else ["GET"]
                
                routes.append({
                    "path": route.path,
                    "methods": methods,
                    "handler": route.endpoint.__name__,
                    "handler_file": handler_file,
                    "handler_line": handler_line,
                    "dependant": route.dependant,
                })
        
        return routes

File: dc-verifier/crates/dc-adapter-fastapi/src/extractor.rs
use anyhow::Result;
use pyo3::prelude::*;
use pyo3::types::PyAny;
use std::path::PathBuf;

/// Извлекает FastAPI приложение и routes
pub struct FastApiExtractor {
    app_path: PathBuf,
}

impl FastApiExtractor {
    /// Создает новый экстрактор
    pub fn new(app_path: PathBuf) -> Self {
        Self { app_path }
    }

    /// Загружает FastAPI app через PyO3
    pub fn load_app(&self) -> Result<Py<PyAny>> {
        Python::attach(|py| {
            // Dynamic import of FastAPI app using importlib
            let importlib = py.import("importlib.util")?;
            let spec_from_file = importlib.getattr("spec_from_file_location")?;
            let module_from_spec = importlib.getattr("module_from_spec")?;

            // Создаем spec из файла
            let app_path_str = self.app_path.to_str().ok_or_else(|| {
                anyhow::anyhow!("App path contains invalid UTF-8: {:?}", self.app_path)
            })?;
            let spec = spec_from_file.call1(("app", app_path_str))?;

            // Получаем loader до перемещения spec
            let loader = spec.getattr("loader")?;
            let module = module_from_spec.call1((spec,))?;

            // Загружаем модуль
            loader.call_method1("exec_module", (module.clone(),))?;

            // Получаем app
            let app = module.getattr("app")?;
            Ok(app.into())
        })
    }

    /// Извлекает routes из FastAPI app
    pub fn extract_routes(&self, app: &Bound<'_, PyAny>) -> Result<Vec<FastApiRoute>> {
        Python::attach(|py| {
            // Получаем routes из app
            let routes = app.getattr("routes")?;
            let routes_list: Vec<Py<PyAny>> = routes.extract()?;

            let mut result = Vec::new();

            // Импортируем APIRoute для проверки типа
            let fastapi_routing = py.import("fastapi.routing")?;
            let api_route_class = fastapi_routing.getattr("APIRoute")?;

            // Импортируем inspect для получения информации о файле
            let inspect = py.import("inspect")?;
            let getfile = inspect.getattr("getfile")?;
            let getsourcelines = inspect.getattr("getsourcelines")?;

            for route in routes_list {
                let route_bound = route.bind(py);

                // Проверяем, является ли это APIRoute
                let is_api_route = route_bound.is_instance(api_route_class.as_ref())?;
                if !is_api_route {
                    continue;
                }

                // Извлекаем path
                let path: String = route_bound.getattr("path")?.extract()?;

                // Извлекаем methods
                let methods_attr = route_bound.getattr("methods")?;
                let methods: Option<Vec<String>> = methods_attr.extract().ok();
                let method = methods
                    .and_then(|m| m.first().cloned())
                    .unwrap_or_else(|| "GET".to_string());

                // Извлекаем endpoint
                let endpoint = route_bound.getattr("endpoint")?;
                let handler: String = endpoint.getattr("__name__")?.extract()?;

                // Получаем информацию о файле и строке
                let (handler_file, handler_line) = match getfile.call1((endpoint.clone(),)) {
                    Ok(file_obj) => {
                        let file_str: String = file_obj.extract()?;
                        let file_path = PathBuf::from(file_str);

                        // Получаем номер строки
                        let line = match getsourcelines.call1((endpoint,)) {
                            Ok(lines_tuple) => {
                                let lines: (Vec<String>, usize) = lines_tuple.extract()?;
                                lines.1
                            }
                            Err(_) => 0,
                        };

                        (file_path, line)
                    }
                    Err(_) => {
                        // Если не удается получить файл, используем app_path
                        (self.app_path.clone(), 0)
                    }
                };

                result.push(FastApiRoute {
                    path,
                    method,
                    handler,
                    handler_file,
                    handler_line,
                });
            }

            Ok(result)
        })
    }
}

/// FastAPI route
#[derive(Debug, Clone)]
pub struct FastApiRoute {
    pub path: String,
    pub method: String,
    pub handler: String,
    pub handler_file: PathBuf,
    pub handler_line: usize,
}
